### PR TITLE
invidious: unstable-2021-11-08 -> unstable-2022-03-16

### PR DESCRIPTION
**Pull request by schnusch**
###### Motivation for this change

Add update script.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/servers/invidious/lsquic.nix
Comment (sbruder) [inline pkgs/servers/invidious/lsquic.nix:15]: I didn’t test this yet, and I don’t know if I’ll have the time to test it in the next days, but AFAIK `lsquic.cr` always requires the lsquic version to match the tag of its version (see [README](https://github.com/iv-org/lsquic.cr#lsquic)). lsquic also requires a specific boringssl version, also specified in [its README](https://github.com/litespeedtech/lsquic#building-boringssl).

Comment (schnusch) [reply pkgs/servers/invidious/lsquic.nix:15]: You are probably right, so the update script should pick the lsquic version from *shard.lock*.

File: pkgs/servers/invidious/update.sh
Comment (sbruder) [inline pkgs/servers/invidious/update.sh:6]: I think this should explicitly change the directory to the root of the package, because otherwise `nix-shell maintainers/scripts/update.nix --argstr package invidious` fails.

```suggestion
cd "$(dirname "${BASH_SOURCE[0]}")"

```

**Review by sbruder — CHANGES_REQUESTED** [review]
Apart from the one thing I noted below, this looks nice and works, thank you!

**Review by sbruder — APPROVED** [review]
Thanks for addressing the feedback, this looks good.

**Comment by schnusch** [comment]
With <https://github.com/iv-org/invidious/pull/2685> `substituteInPlace src/invidious/helpers/helpers.cr ...` needs fixing and other stuffs breaks during build as well.

File: pkgs/servers/invidious/update.sh
Comment (Infinisil) [inline pkgs/servers/invidious/update.sh:5]:
```suggestion
tmp=$(mktemp -d)
trap 'rm -rf "$tmp"' exit
git_dir="$tmp/invidious.git"
```

This way there's a clean new temporary directory every time. Or is it intentional that you're reusing this temporary directory?

File: pkgs/servers/invidious/update.sh
Comment (Infinisil) [inline pkgs/servers/invidious/update.sh:76]: Is this done like this because `lsquic` is not exported from `pkgs.*`? If so, I suggest setting `passthru.lsquic` in the invidious derivation, so that you don't have to hack around this and can use `pkgs.invidious.lsquic.version` instead.

File: pkgs/servers/invidious/update.sh
Comment (Infinisil) [inline pkgs/servers/invidious/update.sh:86]: Oh my, this is really hacky, but it [does appear](https://github.com/litespeedtech/lsquic/pull/346/files) to be how upstream updates their version. We should definitely report this upstream though, that's not good and should be fixed

File: pkgs/servers/invidious/videojs.sh
Comment (Infinisil) [inline pkgs/servers/invidious/videojs.sh:5]: I think a comment mentioning why this is pure would be nice

**Review by infinisil — CHANGES_REQUESTED** [review]
In addition to the below comments, I feel like the update script would be much less hacky if the versions and hashes were stored in JSON files, which are then imported from Nix, instead of Nix directly.

File: pkgs/servers/invidious/update.sh
Comment (Infinisil) [inline pkgs/servers/invidious/update.sh:26]: You can read the `versions.json` file here directly

File: pkgs/servers/invidious/update.sh
Comment (Infinisil) [inline pkgs/servers/invidious/update.sh:68]: Same here

File: pkgs/servers/invidious/update.sh
Comment (Infinisil) [inline pkgs/servers/invidious/update.sh:70]:
```suggestion
        | sed -e 's/-[0-9]*$//')
```

Then you don't need to re-add the `v` prefix in `lsquic.nix`. This should also have a comment explaining why this is needed, which is because [lsquic.cr](https://github.com/iv-org/lsquic.cr)'s versions track [lsquic](https://github.com/litespeedtech/lsquic)'s (see https://github.com/iv-org/lsquic.cr/blob/master/README.md?plain=1#L5), but if there are multiple lsquic.cr releases for the same lsquic one, they seem to add a `-<number>` suffix.

File: pkgs/servers/invidious/update.sh
Comment (schnusch) [reply pkgs/servers/invidious/update.sh:5]: I think [/var/tmp](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch05s15.html) is made for this.

Comment (Infinisil) [reply pkgs/servers/invidious/update.sh:5]: Interesting, I initially thought this was like how `/var/run` is symlinked to `/run` on NixOS, the same doesn't hold for `/var/tmp` and `/tmp` though. I'm thinking that `/var/tmp` is about the system-wide equivalent to the user-local `~/.cache`.

I think I'd probably still prefer the `XDG_CACHE_HOME` with the `~/.cache` fallback since it's configurable and a fairly popular [specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html), but I think `/var/tmp` is fine too.

File: pkgs/servers/invidious/update.sh
Comment (schnusch) [reply pkgs/servers/invidious/update.sh:76]: We must still strip the `v` when we put it in the *versions.json*. I would rather just strip it at the start to keep the variable "clean".

**Comment by isker** [comment]
Thanks for maintaining this branch for so long.  I hope it gets merged soon 🌞.

**Comment by Infinisil** [comment]
Just tested this by cherry-picking this PR, enabling `services.invidious.enable`, going to `localhost:3000` and doing any search. Unfortunately this runs into this issue, which makes it essentially unusable: https://github.com/iv-org/invidious/issues/2753

Title: `SSL_connect: error:1000007d:SSL routines:OPENSSL_internal:CERTIFICATE_VERIFY_FAILED (OpenSSL::SSL::Error)`
Date: `2022-03-12T02:41:05Z`
Route: `/search?q=aoeu`
Version: `2022.03.07-440549f @ master`

<details>
<summary>Backtrace</summary>
<p>
   
```
SSL_connect: error:1000007d:SSL routines:OPENSSL_internal:CERTIFICATE_VERIFY_FAILED (OpenSSL::SSL::Error)
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from ???
```
</p>
</details>

This doesn't happen with the old version without this PR, so this is a regression :/

**Comment by sbruder** [comment]
The reason for this issue is that the statically linked boringssl (required by lsquic) overrides OpenSSL’s CA certificate file location.

It can be fixed either by patching boringssl (like in https://github.com/NixOS/nixpkgs/pull/145882) or by setting the `SSL_CERT_FILE` environment variable (like @squalus pointed out in https://github.com/NixOS/nixpkgs/pull/145882#issuecomment-972104021).

<details><summary>Patching boringssl</summary>

```diff
diff --git a/pkgs/servers/invidious/lsquic.nix b/pkgs/servers/invidious/lsquic.nix
index e44e41cee09..50f6d186705 100644
--- a/pkgs/servers/invidious/lsquic.nix
+++ b/pkgs/servers/invidious/lsquic.nix
@@ -9,6 +9,11 @@ let
       url = "https://boringssl.googlesource.com/boringssl";
       inherit (versions.boringssl) rev sha256;
     };
+
+    patches = (old.patches or [ ]) ++ [
+      # Use /etc/ssl/certs/ca-certificates.crt instead of /etc/ssl/cert.pem
+      ./use-etc-ssl-certs.patch
+    ];
   });
 in
 stdenv.mkDerivation rec {
diff --git a/pkgs/servers/invidious/use-etc-ssl-certs.patch b/pkgs/servers/invidious/use-etc-ssl-certs.patch
new file mode 100644
index 00000000000..ea62378008d
--- /dev/null
+++ b/pkgs/servers/invidious/use-etc-ssl-certs.patch
@@ -0,0 +1,13 @@
+diff --git a/crypto/x509/x509_def.c b/crypto/x509/x509_def.c
+index d2bc3e5c1..329580075 100644
+--- a/crypto/x509/x509_def.c
++++ b/crypto/x509/x509_def.c
+@@ -67,7 +67,7 @@
+
+ #define X509_CERT_AREA          OPENSSLDIR
+ #define X509_CERT_DIR           OPENSSLDIR "/certs"
+-#define X509_CERT_FILE          OPENSSLDIR "/cert.pem"
++#define X509_CERT_FILE          "/etc/ssl/certs/ca-certificates.crt"
+ #define X509_PRIVATE_DIR        OPENSSLDIR "/private"
+ #define X509_CERT_DIR_EVP        "SSL_CERT_DIR"
+ #define X509_CERT_FILE_EVP       "SSL_CERT_FILE"
```
</details>

<details><summary>Setting the SSL_CERT_FILE environment variable with a wrapper</summary>

```diff
diff --git a/pkgs/servers/invidious/default.nix b/pkgs/servers/invidious/default.nix
index 14eed9a3be7..cfe423236f5 100644
--- a/pkgs/servers/invidious/default.nix
+++ b/pkgs/servers/invidious/default.nix
@@ -56,7 +56,7 @@ crystal.buildCrystalPackage rec {
           --replace 'Process.run(%(rsvg-convert' 'Process.run(%(${lib.getBin librsvg}/bin/rsvg-convert'
     '';

-  nativeBuildInputs = [ pkg-config shards ];
+  nativeBuildInputs = [ makeWrapper pkg-config shards ];
   buildInputs = [ libxml2 openssl sqlite ];

   format = "crystal";
@@ -77,6 +77,9 @@ crystal.buildCrystalPackage rec {
   '';

   postInstall = ''
+    wrapProgram $out/bin/invidious \
+        --set SSL_CERT_FILE "/etc/ssl/certs/ca-bundle.crt" \
+
     mkdir -p $out/share/invidious/config

     # Copy static parts
```
</details>

Applying any of those patches fixed the problem for me.

File: pkgs/servers/invidious/default.nix
Comment (sbruder) [inline pkgs/servers/invidious/default.nix:10]: I think this should be changed to reflect that the versions/hashes are now stored in a separate file (and no longer in `lsquic.nix` or `default.nix`).

**Comment by Infinisil** [comment]
The NixOS module should certainly be fixed before this can be merged, the `NIX_SSL_FILE` approach sounds good

**Comment by schnusch** [comment]
sbruder's 970d249d39ccac356c705e6e4db6318ffe93dd42 fixes that.

**Comment by schnusch** [comment]
sbruder's 970d249d39ccac356c705e6e4db6318ffe93dd42 fixes that.

**Review by infinisil — APPROVED** [review]
I think this is looking good! Can you squash down the commits so we only have a single update commit (that matches the PR title), and separate commits for the other changes?

**Comment by schnusch** [comment]
I will squash them before the refactor in c3bae5b1191ddfdc036407884d603cff5cc6717f, otherwise it will get ugly.

**Comment by Infinisil** [comment]
`aarch64-darwin` build fails in CI. Is that expected? If so the platform should be marked as not supported somewhere

**Comment by schnusch** [comment]
The error `Failed to spawn codesign_allocate: No such file or directory` seems fixable, but I don't know the first thing about darwin, so I will mark it as broken.

**Comment by Infinisil** [comment]
Looking good, thanks!

**Pull request merged**
